Timelock tests, natspec and scripts

sajanrajdev opened the pull request:
Added:
- Tests for all timelock and fee recipient operations
- Natpsec for all current ebtc library methods
- Scripts for using all current ebtc library methods with their respective multisigs

Run tests with:
```
brownie test --network sepolia-fork
```

Example of running a script:
```
brownie run scripts/ebtc_treasury.py activePool_claim_fee_recipient_coll_shares 1000 --network sepolia-fork
```

NOTE: You may need to manually setup the Sepolia network, instructions added to Readme.

## Testing results
![image](https://github.com/ebtc-protocol/ebtc-multisig/assets/25463788/a36f6a93-d477-440d-930a-de1f96a4f2d2)

petrovska-petro commented on tests/ebtc/test_timelock_tx_helpers.py:
```diff
@@ -0,0 +1,184 @@
+
+    ## Attempt to schedule grant TechOps minting rights by techopss
+    target = techops.ebtc.authority
+    data = target.setUserRole.encode_input(techops.account, 1, True)
```
One element that still I think will be great to have perhaps in a readme or somewhere a reference of the numerical values/magical numbers of for example for the roles used in `setUserRole`. i.e: what is 2? Or 1? Admin or mint?.

Perhaps creating an enum class within the ebtcClasss may be helpful in aiding into the name of the roles such as ROLE.ADMIN etc

sajanrajdev replied:
Good call. Will add all configuration of roles and signatures to the repo, primarily for reference, but also in a way in which they can be leveraged for scripts and tests.

sajanrajdev replied:
resolved https://github.com/ebtc-protocol/ebtc-multisig/pull/3/commits/f4f448d10451d3a2ea524233cd8ae229b7dd55d1

petrovska-petro commented on scripts/ebtc_governance.py:
```diff
@@ -0,0 +1,220 @@
+    safe.post_safe_tx()
+
+
+def authority_set_public_capability(target_address, functionSig, enabled):
```
Another element that may help while setting role capabilities is to have something similar to the address book, the “signatures” book for signatures such as: mint, burn, set staking reward etc

Basically those which are expected to be used to decrease the probability of failure when calling the scripts, having something like a dictionary where `DICTIONARY_SIGNATURES[“MINT”]` -> ‘0x…..’ may be handy imo

sajanrajdev replied:
resolved https://github.com/ebtc-protocol/ebtc-multisig/pull/3/commits/f4f448d10451d3a2ea524233cd8ae229b7dd55d1

petrovska-petro commented on great_ape_safe/ape_api/ebtc.py:
```diff
@@ -140,11 +167,20 @@ def cancel_lowsec_timelock(
         predecessor=None,
         salt=None,
```
Great job w/ the flexibility of passing either `id` or params into the timelock methods for cancelling

However, will imagine that tracking all those params + unique salt will be tricker at the time of exec the script, so imagine that using only the **id** may be sufficient and the preferred approach, but having the flex is great

how are you envisioning the tracking of the salts used to avoid repeated queue actions crashing on the same id? (i acknowledge that it may have low %, but curious on the approach you have in mind)

sajanrajdev replied:
My idea for tracking all activity on the Timelocks is to build a UI service for displaying all timelock transactions with all their details including id and salts. I am scoping this out but I agree with your concern that it will be a pain to review and manage timelock operations without access to this info.

petrovska-petro replied:
the ui is def a great addition, something that came to my mind that will be valuable to support is that when queuing a tx in the timelock to have a better scope of what storage variables may affect or what modifications represents, would be worth to have a tenderly simulation url to share w/ the rest of potential signers? 

wdyt of adding that feature in the flow while either queuing and/or executing so it prints out on the cli?

petrovska-petro replied:
resolving this comment in light of this ticket https://github.com/ebtc-protocol/ebtc-multisig/issues/16 to be work separaetly so it does not get too dense this pr

petrovska-petro commented on great_ape_safe/ape_api/ebtc.py:
```diff
@@ -282,6 +344,11 @@ def update_timelock_delay(self, new_delay, use_high_sec=False):
     #### ===== CDP MANAGER ===== ####
 
     def cdpManager_set_staking_reward_split(self, value, use_high_sec=False):
```
for this type of txs that are setter that will use `value` will advise to use asserts that are catching early the potential reverts onchain for example in this case:

- setting a value >10k will revert

otherwise may be a "wasted" queue tx imo if nobody catches before hand

petrovska-petro replied:
another though is in the case that some of these values are dynamic based on blockanalytica/treasury/3rd party that modifes them back to the same exact signature due to using the same value, they should be accepting a salt, no?

say, setting from 50 to 60 and then back to 50

sajanrajdev replied:
Good call on the catches, will add.

You are very right on allowing for an optional salt, I missed this. Will add as well and prioritize building a solution for tracking all transactions values.

sajanrajdev replied:
Value catches added: https://github.com/ebtc-protocol/ebtc-multisig/pull/3/commits/157869200187017c7f7fd3be2308e35f1c8c04ac

sajanrajdev replied:
Enabled salts: https://github.com/ebtc-protocol/ebtc-multisig/pull/3/commits/06248d0f4d0d90f313097f580b6ffb94ee2af900

petrovska-petro commented on scripts/ebtc_governance.py:
```diff
@@ -0,0 +1,220 @@
+    safe.post_safe_tx()
+
+
+def cdpManager_set_redemption_fee_floor(value):
```
will have the same suggestion here as for `cdpManager_set_staking_reward_split`. ideally should be the same protective programming for all the methods which are here in `ebtc_governance` and onchain they have an interval that should be expected

sajanrajdev replied:
Addressed https://github.com/ebtc-protocol/ebtc-multisig/commit/157869200187017c7f7fd3be2308e35f1c8c04ac

petrovska-petro commented on great_ape_safe/ape_api/ebtc.py:
```diff
@@ -70,12 +70,26 @@ def __init__(self, safe):
+        @param target The target of the timelock transaction.
+        @param value The ETH value of the timelock transaction.
+        @param data The data of the timelock transaction (encoding of function signature and parameters).
+        @param predecessor The predecessing transacction of the timelock transaction. Matters when cancelling batched transactions.
```
i think that not only matters for "cancelling batched txs", but perhaps it is not it meant purpose afaik

sometimes when posting a tx, we will like to highlighted dependancy between them, which i think its the utility of this variable, no?

meaning before predecessor did not get fully exec, do not allow exec the one im queuing, lmk if im missing something

if the case is like that in which particular cases do you think we may use this @param?

petrovska-petro replied:
i think the commentary for this @param may still be worth to revisit on its description to avoid confusion, since still i think that may not be only a **Matters when cancelling batched transactions**, but also in the scenario that one tx its dependant of another one completion

sajanrajdev replied:
Seems like I had actually corrected all instanced. Current comment;
```
@param predecessor The predecessing transacction of the timelock transaction. Matters when there is a dependency between operations.
```

Reference: https://docs.openzeppelin.com/contracts/4.x/api/governance#timelock-terminology

sajanrajdev commented:
@petrovska-petro, thank you very much for your review, all of your comments have been addressed. Added:
- Mapping of all governance configuration, roles, signatures, actors, etc.. to the eBTC class so that they can be accessed by scripts, tests or just for reference in a maintained place. There are a few changes expected, we should update this as the changes are applied on testnet. Additionally, we should keep this up to date with any changes that occur in the future.
- Added checks before scheduling of any parameter change that is bounded on-chain using the same bounds from the contracts.
- Enabled the use of salts for scheduling duplicate txs.
- Fixed the comments as suggested.
- Wrote new tests for all of the above.

Test suite running properly with new tests and implementations:
![image](https://github.com/ebtc-protocol/ebtc-multisig/assets/25463788/3b9b9265-4aa7-4ee7-8a4d-4fa056176901)
![image](https://github.com/ebtc-protocol/ebtc-multisig/assets/25463788/f2f094ee-0488-4379-a100-8d09cdd14d7b)

petrovska-petro commented:
@sajanrajdev everything looks good and comments were tackled ✅ 

only addition to this pr, which still i believe it has great value while the ui for timelock gets finalised its the side-effect on any timelock tx queue, having in place the tenderly simulation to share to inspect events emitted, storage variable updated, traced etc will be v valuable to share w/ other signers

rayeaster commented on great_ape_safe/ape_api/ebtc.py:
```diff
@@ -177,7 +473,15 @@ def cancel_highsec_timelock(
+        """
+        @dev Grants a role on the timelock to an account.
+        @param role_key The key of the role to grant. Can be one of "PROPOSER_ROLE", "CANCELLER_ROLE", "EXECUTOR_ROLE", or "TIMELOCK_ADMIN_ROLE".
+        @param account The account to grant the role to.
```
seems missing `@param` for `salt`

sajanrajdev replied:
Thanks for the review. Good catch, also noticed this yesterday and addressed on the following PR: https://github.com/ebtc-protocol/ebtc-multisig/pull/17/commits/147550eb643bb5c405e729ecebdb34ca0abdfec7

rayeaster commented on great_ape_safe/ape_api/ebtc.py:
```diff
@@ -198,20 +502,26 @@ def grant_timelock_role(self, role_key, account, use_high_sec=False):
+        """
+        @dev Revokes a role on the timelock from an account.
+        @param role_key The key of the role to revoke. Can be one of "PROPOSER_ROLE", "CANCELLER_ROLE", "EXECUTOR_ROLE", or "TIMELOCK_ADMIN_ROLE".
+        @param account The account to revoke the role from.
```
similarly, missing `@param` for `salt`

rayeaster commented on great_ape_safe/ape_api/ebtc.py:
```diff
@@ -230,47 +540,48 @@ def revoke_timelock_role(self, role_key, account, use_high_sec=False):
+        """
+        @dev Updates the delay on the timelock.
+        @param new_delay The new delay to set on the timelock.
+        @param use_high_sec If true, use the high security timelock. Otherwise, use the low security timelock.
```
again, missing `@param` for `salt`

rayeaster commented on great_ape_safe/ape_api/ebtc.py:
```diff
@@ -281,7 +592,17 @@ def update_timelock_delay(self, new_delay, use_high_sec=False):
+        """
+        @dev Sets the staking reward split in the CDP Manager.
+        @param value The new staking reward split to set.
+        @param use_high_sec If true, use the high security timelock. Otherwise, use the low security timelock.
```
same issue: missing `@param` for `salt`

rayeaster commented on great_ape_safe/ape_api/ebtc.py:
```diff
@@ -290,20 +611,29 @@ def cdpManager_set_staking_reward_split(self, value, use_high_sec=False):
+        """
+        @dev Sets the redemption fee floor in the CDP Manager.
+        @param value The new redemption fee floor to set.
+        @param use_high_sec If true, use the high security timelock. Otherwise, use the low security timelock.
```
same issue: missing `@param` for `salt`

rayeaster commented on great_ape_safe/ape_api/ebtc.py:
```diff
@@ -312,20 +642,29 @@ def cdpManager_set_redemption_fee_floor(self, value, use_high_sec=False):
+        """
+        @dev Sets the minute decay factor for the redemption fee in the CDP Manager.
+        @param value The new redemption fee minute decay factor.
+        @param use_high_sec If true, use the high security timelock. Otherwise, use the low security timelock.
```
same issue: missing `@param` for `salt`

rayeaster commented on great_ape_safe/ape_api/ebtc.py:
```diff
@@ -334,20 +673,23 @@ def cdpManager_set_minute_decay_factor(self, value, use_high_sec=False):
+        """
+        @dev Sets the beta for the redemption fee in the CDP Manager.
+        @param value The new redemption fee beta.
+        @param use_high_sec If true, use the high security timelock. Otherwise, use the low security timelock.
```
same issue: missing `@param` for `salt`

rayeaster commented on great_ape_safe/ape_api/ebtc.py:
```diff
@@ -356,20 +698,25 @@ def cdpManager_set_beta(self, value, use_high_sec=False):
+        """
+        @dev Sets the redemptions paused state in the CDP Manager.
+        @param paused The new redemptions paused state to set (True or False).
+        @param use_high_sec If true, use the high security timelock. Otherwise, use the low security timelock.
```
same issue: missing `@param` for `salt`

rayeaster commented on great_ape_safe/ape_api/ebtc.py:
```diff
@@ -378,20 +725,26 @@ def cdpManager_set_redemptions_paused(self, pause, use_high_sec=False):
+        """
+        @dev Sets the grace period in the CDP Manager.
+        @param value The new grace period to set.
+        @param use_high_sec If true, use the high security timelock. Otherwise, use the low security timelock.
```
same issue: missing `@param` for `salt`

rayeaster commented on great_ape_safe/ape_api/ebtc.py:
```diff
@@ -400,22 +753,27 @@ def cdpManager_set_grace_period(self, value, use_high_sec=False):
+        """
+        @dev Sets the fallbak Oracle caller on the PriceFeed.
+        @param address The address of the new fallback caller
+        @param use_high_sec If true, use the high security timelock. Otherwise, use the low security timelock.
```
same issue: missing `@param` for `salt`

rayeaster commented on great_ape_safe/ape_api/ebtc.py:
```diff
@@ -424,22 +782,28 @@ def priceFeed_set_fallback_caller(self, address, use_high_sec=False):
+        """
+        @dev Sets the fee bps on the Active Pool.
+        @param value The new fee bps.
+        @param use_high_sec If true, use the high security timelock. Otherwise, use the low security timelock.
```
same issue: missing `@param` for `salt`

rayeaster commented on great_ape_safe/ape_api/ebtc.py:
```diff
@@ -448,20 +812,28 @@ def activePool_set_fee_bps(self, value, use_high_sec=False):
+        """
+        @dev Sets the fee bps on the CDP Manager.
+        @param value The new fee bps.
+        @param use_high_sec If true, use the high security timelock. Otherwise, use the low security timelock.
```
same issue: missing `@param` for `salt`

rayeaster commented on great_ape_safe/ape_api/ebtc.py:
```diff
@@ -470,22 +842,30 @@ def borrowerOperations_set_fee_bps(self, value, use_high_sec=False):
+        """
+        @dev Sets the new fee recipient address on the Active Pool.
+        @param address The new fee recipient address.
+        @param use_high_sec If true, use the high security timelock. Otherwise, use the low security timelock.
```
same issue: missing `@param` for `salt`

rayeaster commented on great_ape_safe/ape_api/ebtc.py:
```diff
@@ -494,22 +874,28 @@ def activePool_set_fee_recipient_address(self, address, use_high_sec=False):
+        """
+        @dev Sets the new fee recipient address on the Borrowers Operations.
+        @param address The new fee recipient address.
+        @param use_high_sec If true, use the high security timelock. Otherwise, use the low security timelock.
```
same issue: missing `@param` for `salt`

rayeaster commented on great_ape_safe/ape_api/ebtc.py:
```diff
@@ -518,25 +904,29 @@ def borrowerOperations_set_fee_recipient_addresss(
+        @dev Claims the accumulated collateral shares for the Fee Recipient on the Active Pool.
+        @param value The amount of collateral shares to claim.
+        @param use_timelock If true, use the timelock. Otherwise, use direct tx.
+        @param use_high_sec If true, use the high security timelock. Otherwise, use the low security timelock.
```
same issue: missing `@param` for `salt`

rayeaster commented on great_ape_safe/ape_api/ebtc.py:
```diff
@@ -582,27 +972,37 @@ def activePool_claim_fee_recipient_coll_shares(
+        @dev Sweeps an amount of an specific unprotected, stuck, token from the Active Pool into the fee recipient.
+        @param token_address The address of the token to sweep.
+        @param value The amount of tokens to sweep.
+        @param use_timelock If true, use the timelock. Otherwise, use direct tx.
```
same issue: missing `@param` for `salt`

rayeaster commented on great_ape_safe/ape_api/ebtc.py:
```diff
@@ -613,42 +1013,54 @@ def activePool_sweep_token(
+        @dev Sweeps an amount of an specific unprotected, stuck, token from the Collateral Surplus pool into the fee recipient.
+        @param token_address The address of the token to sweep.
+        @param value The amount of tokens to sweep.
+        @param use_timelock If true, use the timelock. Otherwise, use direct tx.
```
same issue: missing `@param` for `salt`

rayeaster commented on great_ape_safe/ape_api/ebtc.py:
```diff
@@ -613,42 +1013,54 @@ def activePool_sweep_token(
                 fee_recipient = (
-                    self.active_pool.feeRecipientAddress()
+                    self.coll_surplus_pool.feeRecipientAddress()
                 )  # Fee recipient will be modified to track the Active Pool's
```
comment typo? should be `to track Collateral Surplus Pool's` ?

rayeaster commented on great_ape_safe/ape_api/ebtc.py:
```diff
@@ -659,28 +1071,35 @@ def collSurplusPool_sweep_token(
             fee_recipient = (
-                self.active_pool.feeRecipientAddress()
+                self.coll_surplus_pool.feeRecipientAddress()
             )  # Fee recipient will be modified to track the Active Pool's
```
comment typo? should be `to track Collateral Surplus Pool's` ?

rayeaster commented on great_ape_safe/ape_api/ebtc.py:
```diff
@@ -659,28 +1071,35 @@ def collSurplusPool_sweep_token(
+        """
+        @dev Sets the name of a role in the Authority.
+        @param role The role to set the name of.
+        @param name The new name of the role.
```
same issue: missing `@param` for `salt`

rayeaster commented on great_ape_safe/ape_api/ebtc.py:
```diff
@@ -705,12 +1124,18 @@ def authority_set_role_name(self, role, name):
+        @dev Grants a role to a user in the Authority.
+        @param user The user to grant the role to.
+        @param role The role to set the name of.
+        @param enabled Whether to grant or revoke the role.
```
same issue: missing `@param` for `salt`

rayeaster commented on great_ape_safe/ape_api/ebtc.py:
```diff
@@ -731,18 +1156,27 @@ def authority_set_user_role(self, user, role, enabled):
+        @param role The role to set the name of.
+        @param target_address The address of the contract containing the function.
+        @param functionSig The signature of the function to grant the capability to.
+        @param enabled Whether to grant or revoke the role.
```
same issue: missing `@param` for `salt`

rayeaster commented on great_ape_safe/ape_api/ebtc.py:
```diff
@@ -766,18 +1200,26 @@ def authority_set_role_capability(self, role, target_address, functionSig, enabl
+        """
+        @dev Flags a contract's function as callable by anyone (public) in the Authority.
+        @param target_address The address of the contract containing the function.
+        @param functionSig The signature of the function to grant the capability to.
```
same issue: missing `@param` for `salt`

rayeaster commented on great_ape_safe/ape_api/ebtc.py:
```diff
@@ -801,16 +1243,21 @@ def authority_set_public_capability(self, target_address, functionSig, enabled):
+        """
+        @dev Burns the ability to call a contract's function from anyone irrespective of their roles in the Authority.
+        @param target_address The address of the contract containing the function.
+        @param functionSig The signature of the function to grant the capability to.
```
same issue: missing `@param` for `salt`

rayeaster commented on great_ape_safe/ape_api/ebtc.py:
```diff
@@ -833,16 +1280,20 @@ def authority_burn_capability(self, target_address, functionSig):
+    def authority_set_authority(self, new_authority, salt=EmptyBytes32):
+        """
+        @dev Changes the Governance underying authority contract.
+        @param new_authority The address of the new Authority contract.
```
same issue: missing `@param` for `salt`

rayeaster reviewed (approved):
LGTM to have a comprehensive support across eBTC key governance features using timelock facilities

rayeaster reviewed (commented):
LGTM to have missing `@param` for `salt` in a separate commit https://github.com/ebtc-protocol/ebtc-multisig/commit/147550eb643bb5c405e729ecebdb34ca0abdfec7

petrovska-petro reviewed (approved):
reviews were provided from my side few comments above, ready to get it merged ✅
